### PR TITLE
EVG-5636: Restructure Degraded Mode Flags

### PR DESCRIPTION
**Pull request by ybrill**

### CODE REVIEW

File: config_serviceflags.go [hunk 1]
```diff
@@ -8,19 +8,18 @@ import (
...
-	TaskrunnerDisabled           bool `bson:"taskrunner_disabled" json:"taskrunner_disabled"`
-	RepotrackerDisabled          bool `bson:"repotracker_disabled" json:"repotracker_disabled"`
-	SchedulerDisabled            bool `bson:"scheduler_disabled" json:"scheduler_disabled"`
-	GithubPRTestingDisabled      bool `bson:"github_pr_testing_disabled" json:"github_pr_testing_disabled"`
-	RepotrackerPushEventDisabled bool `bson:"repotracker_push_event_disabled" json:"repotracker_push_event_disabled"`
-	CLIUpdatesDisabled           bool `bson:"cli_updates_disabled" json:"cli_updates_disabled"`
-	BackgroundStatsDisabled      bool `bson:"background_stats_disabled" json:"background_stats_disabled"`
-	TaskLoggingDisabled          bool `bson:"task_logging_disabled" json:"task_logging_disabled"`
-	CacheStatsJobDisabled        bool `bson:"cache_stats_job_disabled" json:"cache_stats_job_disabled"`
+	TaskDispatchDisabled    bool `bson:"task_dispatch_disabled" json:"task_dispatch_disabled"`
```
Comment (ybrill) [inline config_serviceflags.go]: whitespace diff

File: public/static/js/admin.js [hunk 2]
```diff
@@ -147,29 +147,6 @@ mciModule.controller('AdminSettingsController', ['$scope', '$window', '$http', '
       });
   }
 
-  var flagDisplayNames = {
```
Comment (ybrill) [inline public/static/js/admin.js]: unused

File: config_serviceflags.go [hunk 3]
```diff
@@ -8,19 +8,18 @@ import (
 
 // ServiceFlags holds the state of each of the runner/API processes
 type ServiceFlags struct {
-	TaskDispatchDisabled         bool `bson:"task_dispatch_disabled" json:"task_dispatch_disabled"`
-	HostinitDisabled             bool `bson:"hostinit_disabled" json:"hostinit_disabled"`
-	MonitorDisabled              bool `bson:"monitor_disabled" json:"monitor_disabled"`
-	AlertsDisabled               bool `bson:"alerts_disabled" json:"alerts_disabled"`
-	TaskrunnerDisabled           bool `bson:"taskrunner_disabled" json:"taskrunner_disabled"`
```
Comment (bsamek) [inline config_serviceflags.go]: We should remove this field from the production and staging databases after deploying this change.

File: config_db.go [hunk 4]
```diff
@@ -53,11 +53,10 @@ var (
 	hostinitKey                     = bsonutil.MustHaveTag(ServiceFlags{}, "HostinitDisabled")
 	monitorKey                      = bsonutil.MustHaveTag(ServiceFlags{}, "MonitorDisabled")
 	alertsKey                       = bsonutil.MustHaveTag(ServiceFlags{}, "AlertsDisabled")
-	taskrunnerKey                   = bsonutil.MustHaveTag(ServiceFlags{}, "TaskrunnerDisabled")
```
Comment (bsamek) [inline config_db.go]: Also change in testdata/smoke_config.yml.

File: config_db.go [hunk 5]
```diff
@@ -53,11 +53,10 @@ var (
...
+	agentStartKey                   = bsonutil.MustHaveTag(ServiceFlags{}, "AgentStartDisabled")
 	repotrackerKey                  = bsonutil.MustHaveTag(ServiceFlags{}, "RepotrackerDisabled")
 	schedulerKey                    = bsonutil.MustHaveTag(ServiceFlags{}, "SchedulerDisabled")
 	githubPRTestingDisabledKey      = bsonutil.MustHaveTag(ServiceFlags{}, "GithubPRTestingDisabled")
-	repotrackerPushEventDisabledKey = bsonutil.MustHaveTag(ServiceFlags{}, "RepotrackerPushEventDisabled")
```
Comment (bsamek) [inline config_db.go]: Also change in testdata/smoke_config.yml.

File: config_db.go (hunk 4 above)
Comment (ybrill) [reply config_db.go]: Is it not changed?

File: config_db.go (hunk 5 above)
Comment (ybrill) [reply config_db.go]: Is it not changed?

File: config_db.go (hunk 4 above)
Comment (bsamek) [reply config_db.go]: Sorry, totally missed it.

File: config_serviceflags.go [hunk 6]
```diff
@@ -8,19 +8,18 @@ import (
...
-	RepotrackerDisabled          bool `bson:"repotracker_disabled" json:"repotracker_disabled"`
-	SchedulerDisabled            bool `bson:"scheduler_disabled" json:"scheduler_disabled"`
-	GithubPRTestingDisabled      bool `bson:"github_pr_testing_disabled" json:"github_pr_testing_disabled"`
-	RepotrackerPushEventDisabled bool `bson:"repotracker_push_event_disabled" json:"repotracker_push_event_disabled"`
-	CLIUpdatesDisabled           bool `bson:"cli_updates_disabled" json:"cli_updates_disabled"`
-	BackgroundStatsDisabled      bool `bson:"background_stats_disabled" json:"background_stats_disabled"`
-	TaskLoggingDisabled          bool `bson:"task_logging_disabled" json:"task_logging_disabled"`
-	CacheStatsJobDisabled        bool `bson:"cache_stats_job_disabled" json:"cache_stats_job_disabled"`
+	TaskDispatchDisabled    bool `bson:"task_dispatch_disabled" json:"task_dispatch_disabled"`
+	HostinitDisabled        bool `bson:"hostinit_disabled" json:"hostinit_disabled"`
```
Comment (tychoish) [inline config_serviceflags.go]: nit: HostInit

File: units/crons.go [hunk 7]
```diff
@@ -113,20 +113,6 @@ func PopulateRepotrackerPollingJobs(part int) amboy.QueueOperation {
...
-
-		if flags.TaskDispatchDisabled {
-			grip.InfoWhen(sometimes.Percent(evergreen.DegradedLoggingPercent), message.Fields{
-				"message": "task dispatching disabled",
-				"mode":    "degraded",
-				"impact":  "skipping stepback activation",
-			})
-			return nil
-		}
-
```
Comment (tychoish) [inline units/crons.go]: why did you remove this?

Comment (bsamek) [reply units/crons.go]: I suggested it. I don't think there's a reason we need to degrade this, as it merely activates versions whose batch times have passed. It doesn't change the size of the queue, and if it were not to run (as part of the scheduler being degraded), it would simply run once we undegraded the scheduler. Since it doesn't put significant load on the system and doesn't affect the queues, I don't think there's a reason to include it in a degraded flag.

File: config_serviceflags.go (hunk 6 above)
Comment (ybrill) [reply config_serviceflags.go]: Should be good now.

File: units/crons.go (hunk 7 above)
Comment (tychoish) [reply units/crons.go]: it can increase the number of tasks that enter the task queues. 

It's weird that it's here and not in the repotracker flag, but I don't think that we should exempt it from degraded mode.

Comment (bsamek) [reply units/crons.go]: > it can increase the number of tasks that enter the task queues.

Sort of. Whether or not the scheduler is degraded determines whether or not tasks enter task queues, and degrading the scheduler stops new tasks from entering the queue altogether. Ultimately the effect of degrading this is merely whether boxes change color from light to dark gray in the UI. We can control the effect of this with other flags, which have clearer effects. That this is part of the repotracker is an historical accident. Batch time isn't logically part of tracking repositories.

Comment (tychoish) [reply units/crons.go]: after a harrowing debate, we agreed that it should be behind the scheduler's flag.

Comment (bsamek) [reply units/crons.go]: @ybrill - The principle we settled on is that more jobs behind degraded flags is better. It doesn't logically fit in the repotracker, since it doens't have to do with talking to github, but it is an example of changing state in a way that ultimately adds things to queues, so the scheduler works here.

Comment (ybrill) [reply units/crons.go]: Done.
